File: plugins/ide-features-trainer/src/org/jetbrains/plugins/feature/suggester/defaultSuggesters/AutoindentLinesSuggester.scala
package org.jetbrains.plugins.feature.suggester.defaultSuggesters

import org.jetbrains.plugins.feature.suggester.{NoSuggestion, Suggestion, FeatureSuggester}
import org.jetbrains.plugins.feature.suggester.changes.{ChildReplacedAction, UserAnAction, UserAction}
import com.intellij.openapi.command.CommandProcessor
import com.intellij.psi.PsiWhiteSpace

/**
 * @author Alefas
 * @since 25.05.13
 */
class AutoindentLinesSuggester extends FeatureSuggester {
  val POPUP_MESSAGE = "Why not to use autoindent feature (Ctrl + Shift + I)"

  def getSuggestion(actions: List[UserAction], anActions: List[UserAnAction]): Suggestion = {
    val command = CommandProcessor.getInstance.getCurrentCommand
    if (command != null) return NoSuggestion
    actions.last match {
      case ChildReplacedAction(parent, child: PsiWhiteSpace, oldChild: PsiWhiteSpace) if child.getText.contains("\n") &&
        child.getText.count(_ == '\n') == oldChild.getText.count(_ == '\n') &&
        child.getText.reverseIterator.takeWhile(_ != '\n').length != oldChild.getText.reverseIterator.takeWhile(_ != '\n').length &&
        child.getNextSibling != null && !child.getNextSibling.isInstanceOf[PsiWhiteSpace] =>
        return SuggestingUtil.createSuggestion(None, POPUP_MESSAGE)
      case _ =>
    }

    NoSuggestion
  }

  def getId: String = "Autoindent lines suggseter"
}

File: plugins/ide-features-trainer/src/main/kotlin/org/jetbrains/plugins/feature/suggester/suggesters/AutoIndentLinesSuggester.kt
package org.jetbrains.plugins.feature.suggester.suggesters

import com.intellij.openapi.command.CommandProcessor
import com.intellij.psi.PsiWhiteSpace
import org.jetbrains.plugins.feature.suggester.FeatureSuggester
import org.jetbrains.plugins.feature.suggester.NoSuggestion
import org.jetbrains.plugins.feature.suggester.Suggestion
import org.jetbrains.plugins.feature.suggester.actions.ChildReplacedAction
import org.jetbrains.plugins.feature.suggester.history.UserActionsHistory

class AutoIndentLinesSuggester : FeatureSuggester {

    companion object {
        const val POPUP_MESSAGE = "Why not use the Auto-Indent feature? (Ctrl + Shift + I)"
    }

    override fun getSuggestion(actions: UserActionsHistory): Suggestion {
        val command = CommandProcessor.getInstance().currentCommand
        if (command != null) return NoSuggestion
        val lastAction = actions.lastOrNull()
        if (lastAction is ChildReplacedAction) {
            val newChild = lastAction.newChild
            val oldChild = lastAction.oldChild
            if (newChild is PsiWhiteSpace && oldChild is PsiWhiteSpace) {
                val newChildText = newChild.text
                val oldChildText = oldChild.text
                if (newChildText.contains('\n')
                    && newChildText.count { it == '\n' } == oldChildText.count { it == '\n' }
                    && newChildText.takeLastWhile { it != '\n' }.length != oldChildText.takeLastWhile { it != '\n' }.length
                    && newChild.nextSibling != null && newChild.nextSibling !is PsiWhiteSpace
                ) {
                    return createSuggestion(null, POPUP_MESSAGE)
                }
            }
        }
        return NoSuggestion
    }

    override fun getId(): String = "AutoIndent lines suggester"
}